### DISCUSSION TITLE
What's the workflow of developing in submodules?

### DISCUSSION THREAD

**Discussion by Raynchowkw** [The-OpenROAD-Project/OpenROAD#683] [General]
I git clone --recursive the parent repo "openroad-flow-scripts" so I get all the code including submodules in the local. 
And then I use "git remote rename origin upstream" and "git remote add origin <my_new_created_blank_openroad-flow.git> to let it point to my repo. (I only create one repo for the openroad-flow-script. I didn't create repo for submodules.)
After I made changes in submodule /tools/openroad/src, I do git `add` and `commit` in the directory of submodule tools/openroad, and then in the parent directory /openroad-flow-scripts, I do `git add tools/openroad`, `commit`, `push`. I can see it does push something to somewhere[![push result](https://i.postimg.cc/jSrTz9Ry/We-Chat-Image-20210526015324.png)](https://postimg.cc/kBwLCT65), but I cannot find it on github.com, on which the directory url points to 404 [![404page](https://i.postimg.cc/FF3DfsHF/404.png)](https://postimg.cc/fkWxpsd1)page. I am wondering where the submodules go when I do the `git remote add origin <my_openroad-flow.git>`. Is there a way to create the submodule repo automatically in my repo?

**Comment by maliberty** [comment]
It isn't clear in which repo you are runing which steps.  Please give a specific set of steps to replay to see your problem.

I think the simplest is to fork ORFS & OR.

**Reply by Raynchowkw** [reply]
What do you mean by fork ORFS & OR? Does it a command name or so? Can I fork recursively?
I temporarily solve it by fork all the submodules and make sure the .gitsubmodule in the parent model points to the right url in my repo.
In my description of my steps, when I say  "in the directory of submodule 'tools/openroad'", it is under the that directory where I execute the commands.  I have only one repo which is "parent directory /openroad-flow-scripts".

**Reply by maliberty** [reply]
I'm a lazy typist ORFS = OpenROAD-flow-scripts; OR = OpenROAD.

The parent repo holds the paths of the submodule in .gitmodules

**Reply by Raynchowkw** [reply]
OK I see. Yeah I have forked them both.
But I still have some questions about the "head attached/detached".
[![git.png](https://i.postimg.cc/VNsTWMkS/git.png)](https://postimg.cc/CRtmFRDY).
After I fork all of them (ORFS, OR, Yosys), I git clone -recursive ORFS to my local. 
I made modifications in OR/src/TritonRoute/src/pa/FlexPA_prep.cpp and use `git add`, `git commit`, 
but I don't know how to push it. What's the meaning of ((2aad13ad...)) and [detached HEAD fbecda7a] here?
How can I add, commit and push it as I do to origin master in my forked repo?

**Comment by maliberty** [comment]
When the submodule is checked out it will be a specific commit that is checked out not the HEAD.  You can just 'it checkout master' and commit your change.  Then in ORFS you can update the submodule to point to your new commit.

**Reply by Raynchowkw** [reply]
Thanks. I have done the git checkout master and commit my change to the master in my OR repo now. How can I let ORFS update the submodule to point to my new commit?

**Comment by maliberty** [comment]
In ORFS you should see that the submodule is modified and you git add/commit it as usual.

**Reply by Raynchowkw** [reply]
OK. In git bash, I go to the directory of /ORFS, and use `git status` to see that there is modification in /tools/OR. And then I `git add tools/OR`, `git commit`, `git push`. Is it correct?

**Comment by maliberty** [comment]
yes

**Reply by Raynchowkw** [reply]
Thanks.

**Comment by Raynchowkw** [comment]
When I want to get the update of submodules, I do `git pull --recurse-submodule`, [![gitpullsubmodulecheckout.png](https://i.postimg.cc/hPpJ5z1G/gitpullsubmodulecheckout.png)](https://postimg.cc/svGDB28F),
but it checkouts the commit id it originally points to. [![submodulepointtocommit.png](https://i.postimg.cc/5tF6yPdH/submodulepointtocommit.png)](https://postimg.cc/Y4MpyzSk).
How to let it point to my OR git master branch? Otherwise, I need to go into that submodule directory to do the `git checkout master` every time.

**Reply by rovinski** [reply]
Submodules do not track branches. They track commits. Every time you make a commit into a submodule, you would have to update the parent repository. An example workflow would be:

1. Navigate to the submodule OpenROAD: `cd tools/OpenROAD/`
2. Modify a file (such as `src/TritonRoute/src/TritonRoute.tcl`)
3. Commit the change to the submodule (while your current directory is inside `tools/OpenROAD/`): `git add src/TritonRoute/src/TritonRoute.tcl; git commit`.
4. Push changes to the submodule (while your current directory is inside `tools/OpenROAD/`): `git push`
5. Navigate to the parent repository: `cd ../..`
6. Commit the change to the submodule: `git add tools/OpenROAD/; git commit`
7. Push the change to the parent repository: `git push`

**Reply by Raynchowkw** [reply]
Thanks for your detailed explanation!
